2019-08-time-series/stochastic_volatility.py

```diff
@@ -0,0 +1,182 @@
+from pyro.infer.mcmc.util import predictive, initialize_model
+from pyro.optim import Adam
+
+logging.basicConfig(format='%(relativeCreated) 9d %(message)s', level=logging.INFO)
+
+
+"""
+Multivariate Stochastic Volatility Model
+See Model 2 on page 365 in https://pdfs.semanticscholar.org/fccc/6f4ee933d4330eabf377c08f8b2650e1f244.pdf
+"""
+def model(data):
+    """
+    y = diag(exp(h_t / 2)) * eps_t
+    eps ~ Q L_eps rho_t
+    h_{i+1} = mu + Phi(h_t - mu) + eta
+
+    We do this in log space to convert multiplicative noise to additive noise
+    so we can leverage the GaussianHMM distribution.
+
+    log y_kt = h_kt / 2 + log <L_eps, delta_t>
+            ~= h_kt / 2 + gamma_kt where gamma ~ MVN(0. sigma)
+    and we moment match to compute epsilon.
+
+    :param data: Tensor of the shape ``(securities, timesteps, returns)``
+    :type data: torch.Tensor
+    """
+    obs_dim = data.shape[-1]
+    hidden_dim = obs_dim
+    with pyro.plate(len(data)):
+        mu = pyro.param('mu', torch.zeros(hidden_dim))
+        L = pyro.param('L', torch.eye(hidden_dim), constraint=constraints.lower_cholesky)
+        init_dist = dist.MultivariateNormal(mu, scale_tril=L)
+
+        L_eta = pyro.param('L_eta', torch.eye(hidden_dim), constraint=constraints.lower_cholesky)
+        mu_eta = torch.zeros(hidden_dim)
+        trans_matrix = pyro.param('phi', torch.ones(hidden_dim))
+        # this gives us a zero matrix with phi on the diagonal
+        trans_matrix = trans_matrix.diag()
+        trans_dist = dist.MultivariateNormal(mu_eta, scale_tril=L_eta)
+
+        mu_gamma  = pyro.param('mu_gamma', torch.zeros(obs_dim))
+        L_gamma = pyro.param('sigma_gamma', torch.eye(obs_dim), constraint=constraints.lower_cholesky)
+        obs_matrix = torch.eye(hidden_dim, obs_dim)
+        # latent state is h_t - mu
+        obs_dist = dist.MultivariateNormal(-mu_gamma, scale_tril=L_gamma)
+
+        pyro.sample('obs', dist.GaussianHMM(init_dist, trans_matrix, trans_dist, obs_matrix, obs_dist), obs=data)
+
+
+def hmc_model(data):
+    """
+    y = diag(exp(h_t / 2)) * eps_t
+    eps ~ Q L_eps rho_t
+    h_{i+1} = mu + Phi(h_t - mu) + eta
+
+    We do this in log space to convert multiplicative noise to additive noise
+    so we can leverage the GaussianHMM distribution.
+
+    log y_kt = h_kt / 2 + log <L_eps, delta_t>
+            ~= h_kt / 2 + gamma_kt where gamma ~ MVN(0. sigma)
+    and we moment match to compute epsilon.
+
+    :param data: Tensor of the shape ``(securities, timesteps, returns)``
+    :type data: torch.Tensor
+    """
+    one = torch.ones(1)
+    hidden_dim = 4
+    obs_dim = data.shape[-1]
+    with pyro.plate(len(data)):
+        mu = pyro.sample('mu', dist.Normal(torch.zeros(hidden_dim), torch.ones(hidden_dim)))
+        theta = pyro.sample("theta", dist.HalfCauchy(torch.ones(hidden_dim)))
+        L = pyro.sample('L', dist.LKJCorrCholesky(hidden_dim, one))
+        L = torch.mm(torch.diag(theta.sqrt()), L)
```

Comment:
Another possibility is to replace these MVN priors by LowRankMVN
```
cov_factor = pyro.sample('cov_factor', dist.Normal(torch.zeros(hidden_dim, 1), one))  # rank 1
cov_diag = pyro.sample('cov_diag', dist.HalfCauchy(torch.ones(hidden_dim)))
init_dist = dist.LowRankMultivariateNormal(mu, cov_factor=cov_factor, cov_diag=cov_diag)
```
which might be better for applying [mvn_to_gaussian](https://github.com/pyro-ppl/pyro/blob/dev/pyro/ops/gaussian.py#L253) because assuming cov_factor is N x k, to get precision matrix, only the cholesky of k x k matrix is required. I expected that with LowRankMVN, potential_fn is jittable so MCMC might be faster. But when using it, time to run MCMC for this model is only reduced from 30m to 26m. :(

Reply:
thanks for reviewing - no need to try to optimize this as i'm probably going to remove the hmc implementation in the short term, the proper solution seems to be to use numpyro ;)